lib/src/common/shared.rs

```diff
@@ -164,15 +163,26 @@ impl<T: Grounded> Grounded for Shared<T> {
         rust_type_atom::<Self>()
     }
 
-    fn match_(&self, other: &Atom) -> MatchResultIter {
-        self.0.borrow().match_(other)
+    fn as_match(&self) -> Option<&dyn CustomMatch> {
+        match self.0.borrow().as_match() {
+            None => None,
+            Some(_) => Some(self),
+        }
     }
 
     fn execute(&self, args: &[Atom]) -> Result<Vec<Atom>, ExecError> {
         self.0.borrow().execute(args)
     }
 }
 
+impl<T: Grounded> CustomMatch for Shared<T> {
```

Comment:
I don't love blanket impls for something like this.  But my issue is with the Shared type more generally, so I think the code in this PR is fine in the context of the rest of the design.

Reply:
I am not sure what would be a better solution here. Do you mean we should require from user manually implementing `Grounded` and `CustomMatch` for `Shared<SomeType>` if user wants using `Shared` reference as a grounded atom?

Reply:
My issue is with Shared in general (not with this PR). But the internal mutability pattern (RefCell) comes with big implications for how you should access the underlying objects.  Therefore the pattern should be incorporated into the object's API itself instead of creating a wrapper that can wrap another object.

Otherwise it leads to bugs like https://github.com/trueagi-io/hyperon-experimental/issues/241 (as just one example)

So if a grounded atom needs to be a reference to a larger structure with shared mutability, (like with a Space) creating a reference object (e.g. DynSpace) that manages access is a better solution.